providers/siw/siw.h

```diff
@@ -0,0 +1,100 @@
+/* SPDX-License-Identifier: GPL-2.0 or BSD-3-Clause */
+
+/* Authors: Bernard Metzler <bmt@zurich.ibm.com> */
+/* Copyright (c) 2008-2019, IBM Corporation */
+
+#ifndef _SIW_H
+#define _SIW_H
+
+#include <pthread.h>
+#include <inttypes.h>
+#include <stddef.h>
+
+#include <infiniband/driver.h>
+#include <infiniband/kern-abi.h>
+
+struct siw_device {
+	struct verbs_device base_dev;
+};
+
+struct siw_srq {
+	struct ibv_srq base_srq;
+	struct siw_rqe *recvq;
+	uint32_t rq_put;
+	uint32_t num_rqe;
+	pthread_spinlock_t lock;
+};
+
+struct siw_mr {
+	struct verbs_mr base_mr;
+};
+
+struct siw_qp {
+	struct ibv_qp base_qp;
+	struct siw_device *siw_dev;
+
+	uint32_t id;
+
+	pthread_spinlock_t sq_lock;
+	pthread_spinlock_t rq_lock;
+
+	struct ibv_post_send db_req;
+	struct ib_uverbs_post_send_resp db_resp;
+
+	uint32_t num_sqe;
+	uint32_t sq_put;
+	int sq_sig_all;
+	struct siw_sqe *sendq;
+
+	uint32_t num_rqe;
+	uint32_t rq_put;
+	struct siw_rqe *recvq;
+	struct siw_srq *srq;
+};
+
+struct siw_cq {
+	struct ibv_cq base_cq;
+	struct siw_device *siw_dev;
+	uint32_t id;
+
+	/* Points to kernel shared control
+	 * object at the end of CQE array
+	 */
+	struct siw_cq_ctrl *ctrl;
+
+	int num_cqe;
+	uint32_t cq_get;
+	struct siw_cqe *queue;
+	pthread_spinlock_t lock;
+};
+
+struct siw_context {
+	struct verbs_context base_ctx;
+	uint32_t dev_id;
+};
+
+#undef offsetof
+#ifdef __compiler_offsetof
+#define offsetof(TYPE, MEMBER) __compiler_offsetof(TYPE, MEMBER)
+#else
+#define offsetof(TYPE, MEMBER) ((size_t)&((TYPE *)0)->MEMBER)
+#endif
+
+#ifndef container_of
+#define container_of(ptr, type, member)                                        \
```

Comment:
Use the builtin container_of, it is better than this

Reply:
right. this is legacy stuff where container_of was not there in all compilers I used...

will throw that all away